### DISCUSSION TITLE
How to handle Unicode command arguments on Windows?

### DISCUSSION THREAD

**Discussion by YunchengLiu** [daniele77/cli#179] [Q&A]
Hi, thanks for this great library.

I want to know if this lib will handle unicode data that input from windows command?

I simply bound the function of the following interface, and found that as long as the input path contains UNICODE (Chinese), the content of the path cannot be parsed correctly.

``` c++
  void conv_func(std::ostream& os, std::string const& path) {

      // read content from path
      // impl....

  }
```

I'm wondering, what is the best practice to handle this situation?

I use MSVC (C++20), compiling on Windows.

Or can I use `std::wstring` as an interface function parameter to adapt to the default wide character under Windows?

**Comment by daniele77** [comment]
Actually, I never used it with Unicode.
Yes, try with `std::wstring` as parameter, then please let me know if it works.
Thanks.

**Comment by YunchengLiu** [comment]
Sorry for this late reply.

 `std::wstring` seems not to be ok, becaule this library's implement is denpent on the narrow char container (`std::string`).  So if just use `std::wstring`, a compile error generated.

I think the easiest way is to provide a unicode-friendly input and output stream object as session parameter to replace `std::cin` and `std::cout`. So this library do not need to handle input content's encoding from Windows console, 

If do not mind introducing 3rd lib,  [boost::nowide](https://www.boost.org/doc/libs/1_81_0/libs/nowide/doc/html/index.html) is good portable way.

Since parameters is in utf-8 encoding, uncode data is ok, and processing or converting it becomes eaiser.

```c++
cli::Cli app{/*init code*/};

// FileSession session{app}; // old

// now, all parameters is in a std::string encoded UTF-8
FileSession session{app, boost::nowide::cin, boost::nowide::cout};

/*....*/

void a_command_fun(std::ostream& os, std::string const& p){
 // p is in UTF-8 encoding, unicode data is in it
 // just to use or convert it
}

```

**Reply by daniele77** [reply]
Can `boost::nowide` be used only in the client-side code, or does the cli library need to support it?
From the example code you posted, it's enough to pass boost::nowide parameters to the library.

**Comment by YunchengLiu** [comment]
I think use `boost::nowide` in the client-side code is out-of-box enough, `cli` has enough customization in terms of input and output streams (at least for the part I use, pass custom stream object as parameters is enough).

I recommand  this point can be mentioned in the document.

**If there is a need to operate UNICODE through input and output streams, a suitable custom stream object should be provided to replace the default. For example, consider using cross-platform `boost:: nowide` to handle Unicode aware i/o stream** .

**Reply by daniele77** [reply]
Great. If we only need to add a sentence to the README, let's do it!
Can you do a PR?

**Comment by YunchengLiu** [comment]
I have created a PR https://github.com/daniele77/cli/pull/187 for this.

**Reply by daniele77** [reply]
Merged. Thank you very much.
